src/test/java/org/broadinstitute/hellbender/utils/locusiterator/LocusIteratorByStateUnitTest.java

```diff
@@ -0,0 +1,745 @@
+
+            tester.stepForwardOnGenome();
+
+            if ( ! hasNeighboringPaddedOps(params.getElements(), pe.getCurrentCigarOffset()) ) {
+                Assert.assertEquals(pe.isBeforeDeletionStart(), tester.isBeforeDeletionStart, "before deletion start failure");
+                Assert.assertEquals(pe.isAfterDeletionEnd(), tester.isAfterDeletionEnd, "after deletion end failure");
+            }
+
+            Assert.assertEquals(pe.isBeforeInsertion(), tester.isBeforeInsertion, "before insertion failure");
+            Assert.assertEquals(pe.isAfterInsertion(), tester.isAfterInsertion, "after insertion failure");
+            Assert.assertEquals(pe.isNextToSoftClip(), tester.isNextToSoftClip, "next to soft clip failure");
+
+            Assert.assertTrue(pe.getOffset() >= lastOffset, "Somehow read offsets are decreasing: lastOffset " + lastOffset + " current " + pe.getOffset());
+            Assert.assertEquals(pe.getOffset(), tester.getCurrentReadOffset(), "Read offsets are wrong at " + bpVisited);
+
+            Assert.assertEquals(pe.getCurrentCigarElement(), read.getCigar().getCigarElement(tester.currentOperatorIndex), "CigarElement index failure");
+            Assert.assertEquals(pe.getOffsetInCurrentCigar(), tester.getCurrentPositionOnOperatorBase0(), "CigarElement index failure");
+
+            Assert.assertEquals(read.getCigar().getCigarElement(pe.getCurrentCigarOffset()), pe.getCurrentCigarElement(), "Current cigar element isn't what we'd get from the read itself");
+
+            Assert.assertTrue(pe.getOffsetInCurrentCigar() >= 0, "Offset into current cigar too small");
+            Assert.assertTrue(pe.getOffsetInCurrentCigar() < pe.getCurrentCigarElement().getLength(), "Offset into current cigar too big");
+
+            Assert.assertEquals(pe.getOffset(), tester.getCurrentReadOffset(), "Read offset failure");
+            lastOffset = pe.getOffset();
+        }
+
+        final int expectedBpToVisit = read.getEnd() - read.getStart() + 1;
+        Assert.assertEquals(bpVisited, expectedBpToVisit, "Didn't visit the expected number of bp");
+    }
+
+    // ------------------------------------------------------------
+    //
+    // Tests for keeping reads
+    //
+    // ------------------------------------------------------------
+
+    @DataProvider(name = "LIBS_ComplexPileupTests")
+    public Object[][] makeLIBS_ComplexPileupTests() {
+        final List<Object[]> tests = new LinkedList<>();
+
+        for ( final int downsampleTo : Arrays.asList(-1, 1, 2, 5, 10, 30)) {
+            for ( final int nReadsPerLocus : Arrays.asList(1, 10, 60) ) {
+                for ( final int nLoci : Arrays.asList(1, 10, 25) ) {
+                    for ( final int nSamples : Arrays.asList(1) ) {
+                        for ( final boolean keepReads : Arrays.asList(true, false) ) {
+                            for ( final boolean grabReadsAfterEachCycle : Arrays.asList(true, false) ) {
```

Comment:
Can you please confirm that this combinatorial testing is not causing inappropriate runtime for this test class?